python/louis/__init__.py.in

```diff
@@ -327,11 +354,78 @@ def getTypeformForEmphClass(tableList, emphClass):
     tablesString = _createTablesString(tableList)
     return liblouis.lou_getTypeformForEmphClass(tablesString, emphClass)
 
+def dotsToChar(tableList, inbuf):
+    """"Convert a string of dot patterns to a string of characters according to the specifications in tableList.
+    @param tableList: A list of translation tables.
+    @type tableList: list of str
+    @param inbuf: a string of dot patterns, either in liblouis format or Unicode braille.
+    @type inbuf: str
+    @raise RuntimeError: If a complete conversion could not be done.
+    @see: lou_dotsToChar in the liblouis documentation
+    """
+    tablesString = _createTablesString(tableList)
+    inbuf = createEncodedByeString(inbuf)
+    length = c_int(len(inbuf) // wideCharBytes)
+    outbuf = create_string_buffer(length.value * wideCharBytes)
+    if not liblouis.lou_dotsToChar(tablesString, inbuf, outbuf, length, 0):
+        raise RuntimeError("Can't convert dots to char: tables %s, inbuf %s"%(tableList, inbuf))
+    return outbuf.raw[:length.value * wideCharBytes].decode(conversionEncoding)
+
+def charToDots(tableList, inbuf, mode=0):
+    """"Convert a string of characterss to a string of dot patterns according to the specifications in tableList.
+    @param tableList: A list of translation tables.
+    @type tableList: list of str
+    @param inbuf: a string of characters.
+    @type inbuf: str
+    @param mode: The translation mode; add multiple values for a combined mode.
+    @type mode: int
+    @raise RuntimeError: If a complete conversion could not be done.
+    @see: lou_charToDOts in the liblouis documentation
+    """
+    tablesString = _createTablesString(tableList)
+    inbuf = createEncodedByeString(inbuf)
+    length = c_int(len(inbuf) // wideCharBytes)
+    outbuf = create_string_buffer(length.value * wideCharBytes)
+    if not liblouis.lou_charToDots(tablesString, inbuf, outbuf, length, mode):
+        raise RuntimeError("Can't convert char to dots: tables %s, inbuf %s, mode %d"%(tableList, inbuf, mode))
+    return outbuf.raw[:length.value * wideCharBytes].decode(conversionEncoding)
+
+def registerLogCallback(logcallback):
+    """Register logging callbacks.
+    Set to C{None} for default callback.
+    @param logcallback: The callback to use.
+        The callback must take two arguments:
+        @param level: The log level on which a message is logged.
```

Comment:
Too many tabs.

Reply:
This is an argument for the callback, not for the outer function. Not sure whether this is valid though.